examples/text_generation/tutorial_generate_text.py

```diff
@@ -222,166 +243,88 @@ def main_lstm_generate_text():
-    cost = loss_fn(network, targets, batch_size, sequence_length)
-
-    # Truncated Backpropagation for training
-    with tf.variable_scope('learning_rate'):
-        lr = tf.Variable(0.0, trainable=False)
-    # You can get all trainable parameters as follow.
-    # tvars = tf.trainable_variables()
-    # Alternatively, you can specify the parameters for training as follw.
-    #  tvars = network.all_params      $ all parameters
-    #  tvars = network.all_params[1:]  $ parameters except embedding matrix
-    # Train the whole network.
-    tvars = rnn_model.trainable_weights
-    # grads, _ = tf.clip_by_global_norm(tf.gradients(cost, tvars), max_grad_norm)
-    # optimizer = tf.train.GradientDescentOptimizer(lr)
-    train_op = tf.train.GradientDescentOptimizer(lr).minimize(cost, var_list=tvars)
-    # grads = tf.gradients(cost, tvars)
-    # grads[0] = tf.convert_to_tensor(grads[0].values)
-    # grads = optimizer.compute_gradients(cost, tvars)
-    # grads[0][0] = tf.convert_to_tensor(grads[0][0].values)
-    # grads, _ = tf.clip_by_global_norm(grads, max_grad_norm)
-    # train_op = optimizer.apply_gradients(zip(grads, tvars))
+    init = tl.initializers.random_uniform(-init_scale, init_scale)
+
+    net = Text_Generation_Net(vocab_size, hidden_size, init)
+
+    train_weights = net.weights
+    optimizer = tf.optimizers.Adam(lr=learning_rate)
 
     # ===== Training
-    sess.run(tf.global_variables_initializer())
 
     print("\nStart learning a model to generate text")
-    for i in range(max_max_epoch):
-        # decrease the learning_rate after ``max_epoch``, by multipling lr_decay.
-        new_lr_decay = lr_decay**max(i - max_epoch, 0.0)
-        sess.run(tf.assign(lr, learning_rate * new_lr_decay))
+    for i in range(max_epoch):
 
-        print("Epoch: %d/%d Learning rate: %.8f" % (i + 1, max_max_epoch, sess.run(lr)))
+        print("Epoch: %d/%d" % (i + 1, max_epoch))
         epoch_size = ((len(train_data) // batch_size) - 1) // sequence_length
 
         start_time = time.time()
         costs = 0.0
         iters = 0
+
+        net.train()
         # reset all states at the begining of every epoch
-        state1 = tl.layers.initialize_rnn_state(lstm1.initial_state)
+        lstm_state = None
         for step, (x, y) in enumerate(tl.iterate.ptb_iterator(train_data, batch_size, sequence_length)):
-            _cost, state1, _ = sess.run(
-                [cost, lstm1.final_state, train_op], feed_dict={
-                    input_data: x,
-                    targets: y,
-                    lstm1.initial_state: state1
-                }
-            )
-            costs += _cost
-            iters += sequence_length
+            with tf.GradientTape() as tape:
+                ## compute outputs
+                logits, lstm_state = net(x, initial_state=lstm_state)
+                ## compute loss and update model
+                cost = tl.cost.cross_entropy(logits, tf.reshape(y, [-1]), name='train_loss')
```

Comment:
@JingqingZ it is the correct loss here??

Reply:
also, I saw the `max_max_epoch` is deleted, not sure whether is it the reason that @1FengL 's loss is different from the loss of the original implementation?

Reply:
Hi, the max_max_epoch is deleted in text generation tutorial for simplicity, should I add it back? The loss of text generation tutorial is fine and it is the ptb tutorial whose loss is abnormal.